### ISSUE TITLE
RSSFeed Scanner occasionally adds "Metadata download for..." jobs in STOPPED state

### ISSUE THREAD

**Issue by ferdnyc** [BiglySoftware/BiglyBT#2777]
I don't know if it's the fault of RSSFeed Scanner (possibly only since its recent new version update), or it's caused by the BiglyBT core itself, but recently I'm seeing _occasional_ instances of feed magnet URL downloads being created in "STOPPED" state -- meaning, the _metadata_ download job is stopped, rather than the final torrent job. (I have feed downloads configured to be added in STOPPED state, so that part's expected. But BiglyBT's supposed to actually download the torrent first!)

There doesn't appear to be any pattern to why or when this happens, and it occurs in well under 10% of torrents added by the plugin. But it's now happened several times over the past week, already.

I can get the actual torrent successfully loaded if I delete the "Metadata download..." entry, then double-click the corresponding row in RSSFeed Scanner's list to re-get the magnet link.

If I try to manually **start** the original stopped "Metadata download..." job, it never downloads anything and just sits there at "ETA: ∞" forever.

### System block

> Java 11.0.16.1 (64 bit)
>   Red Hat, Inc.
> /usr/lib/jvm/java-11-openjdk-11.0.16.1.1-1.fc36.x86_64
> 
> SWT v4956r12, gtk/3.24.34, zoom=100, dpi=96
> Linux v5.19.13-200.fc36.x86_64, amd64 (64 bit)
> B3.2.0.1_B07/4 az3

RSSFeed Scanner is the recent/current 1.8.3 version.

**Comment by ferdnyc** [comment]
(I should also clarify that I don't know for _certain_ that the metadata download jobs are **added** in STOPPED state. I haven't seen one being added _at all_, I just find them in the Library that way. So it's theoretically _possible_, though it seems unlikely, that they were added in FORCED DOWNLOAD state like normal, and somehow _became_ STOPPED after the fact.)

**Comment by parg** [comment]
Give B08 a whirl, I fiddled with a few things

**Comment by ferdnyc** [comment]
Updating now, we'll see how it goes. I see that I currently have another one of those waiting for me from last night's feed pickups, so good timing! (Determining whether B08 actually fixes it might take a few days, tho. I'll need to wait for more torrents to hit the feeds.)

**Comment by ferdnyc** [comment]
Huh. Things... might be _worse_, actually. Well... maybe not, but I think I do at least have a better handle on the bug now. (Not fixed completely, though.) Here's what I saw:

When BiglyBT restarted, the formerly-STOPPED "Metadata download for..." entry had disappeared from the Library view (as expected), so I went over to the RSSFeed Scanner list, found it, and double-clicked to download it. 

Then, as I sat there watching, a new, replacement "Metadata download for..." entry appear in my Library, in FORCED DOWNLOAD state, _started_ downloading from one of the peers, and then seemed to suddenly shift itself into STOPPED state for no reason at all.

...But here's the funny thing: It DID download the torrent file successfully, and has the path listed in the "Torrent Location:" field of its Info view.

So it seems like the only issue is that, when the torrent finishes downloading, the import of the actual torrent (in STOPPED state, as I have it configured) to replace the metadata job isn't happening properly.  Instead, the metadata job stays in the Library, and _it_ goes into STOPPED state. But only after it's actually finished the download. (Presumably the torrents had been successfully downloaded in all of the other instances of this too, and I just didn't realize.)

**Comment by ferdnyc** [comment]
Hmm. I wanted to wait for another of these to happen on its own, to see if the logs caught anything. One did, and they might have. Mostly it's just the standard networking logspam, but this is unusual, and at least vaguely falls within the timeframe of the occurrence:

> DEBUG::Tue Nov 08 00:28:48 EST 2022::com.biglybt.core.util.DirectByteBufferPoolReal::getSliceBuffer::816:
>   Run out of slice space for '2048, reverting to normal allocation
>     getBufferHelper (DirectByteBufferPoolReal.java:262), getBufferSupport (DirectByteBufferPoolReal.java:243), getBuffer (DirectByteBufferPool.java:49), write (TransportHelperFilterStream.java:199), write (TransportHelperFilterInserter.java:52), write (TransportImpl.java:299), deliverToTransport (OutgoingMessageQueueImpl.java:630), write (MultiPeerUploader.java:355), doProcessing (MultiPeerUploader.java:572), doHighPriorityWrite (WriteController.java:459), writeProcessorLoop (WriteController.java:292), runSupport (WriteController.java:111), run (AEThread.java:72)
>

**Comment by parg** [comment]
Ignore that log-spew, unrelated!

The way magnet downloads work is, as always, steeped in history.

Initially Azureus had its own way of downloading the torrent files that only used the DHT. This has some benefits - for example it works when other users have stopped/queued torrents, doesn't require an actual download to be added to the library in a strange state (e.g. its size isn't known). 

Then bittorrent created their own way based on the ut_metadata extension requiring peer connections etc.

Eventually Azureus supported both mechanisms in parallel - when a magnet is added the DHT approach is used in parallel with the actual download. The actual metadata download is added after a (configurable) time delay so if there are Azureus/Vuze/BiglyBT peers out there with the torrent is is possible for the process to complete without even creating the download.

Once the metadata download completes (or needs discarding) it is removed from BiglyBT and the actual download added in its place. You have to remember that when this was implemented there were lots of plugins out there (some private) that wouldn't take nicely to the metadata download morphiing into a real one with a different size, trackers, who knows what.

Internally BiglyBT identifies downloads by their torrent hash. The metadata download and the actual download have the same hash. There is a potential race between the removal of the metadata download and the addition of the real one. The recent changes I have made attempt to close some windows where confusion could occur.

I have tried for a long time to reproduce your "metadata download in stopped state" problem with RSS Feed have haven't managed it. Found a few other issues but am a bit stumped.

I'll add some logging to the DiskOps log file as it is a handy location in the next beta + plugin version

**Comment by ferdnyc** [comment]
> I have tried for a long time to reproduce your "metadata download in stopped state" problem with RSS Feed have haven't managed it. Found a few other issues but am a bit stumped.

I'm actually not so certain this has anything to do with RSSFeed Scanner, anymore. Since the magnet downloads it adds **start out** completely normal (I've now seen), and only "break" in the process of transitioning from metadata download to actual download, can RSSFeed even be involved in that part of the process, to be breaking it?

I'm also not certain this _wouldn't_ also happen if I were to load the magnets via the clipboard, or some other method. (I haven't _seen_ that  happen, but I also hardly ever manually import magnet links, period.)

**Comment by ferdnyc** [comment]
👍 Installing now...

**Comment by ferdnyc** [comment]
OK, sorry that took a while. Some of the log messages showed torrent names, some showed hashes, so I had to match them up to each other and then sanitize the names.

There does appear to be a pattern: Out of five feed entries loaded, the two that failed to complete the transition to regular download both showed `existingIsMagnet=true` in the RSSFeed log messages. That did not appear on any of the ones that successfully became regular downloads.

Sanitized partial logs:
```text
[12:30:38] RSSFeed: adding without options: magnet returned: false
[12:30:39] RSSFeed: adding without options: HASH_1
[12:30:39] addDownload: HASH_1: isMagnet=false
[12:30:39] RSSFeed: adding without options: magnet returned: false
[12:31:36] addDownload: HASH_2: isMagnet=true
[12:31:38] addDownload: HASH_3: isMagnet=true
[12:31:38] addDownload: HASH_5: isMagnet=true
[12:31:40] addDownload: HASH_4: isMagnet=true
[12:31:42] addDownload: HASH_6: isMagnet=true
[12:31:49] RSSFeed: adding without options: HASH_5
[12:31:49] addDownload: HASH_5: isMagnet=false, existingIsMagnet=true
[12:31:49] "Metadata download for HASH_5.torrent" is not persistent, auto-move disabled.
[12:31:49] RSSFeed: adding without options: magnet returned: true
[12:31:49] RSSFeed: adding without options: HASH_2
[12:31:49] addDownload: HASH_2: isMagnet=false, existingIsMagnet=true
[12:31:49] "Metadata download for HASH_2.torrent" is not persistent, auto-move disabled.
[12:31:49] RSSFeed: adding without options: magnet returned: true
[12:31:52] "Metadata download for HASH_3.torrent" is not persistent, auto-move disabled.
[12:31:52] RSSFeed: adding without options: HASH_3
[12:31:52] addDownload: HASH_3: isMagnet=false
[12:31:52] RSSFeed: adding without options: magnet returned: false
[12:32:03] "Metadata download for HASH_4.torrent" is not persistent, auto-move disabled.
[12:32:03] RSSFeed: adding without options: HASH_4
[12:32:03] addDownload: HASH_4: isMagnet=false
[12:32:03] RSSFeed: adding without options: magnet returned: false
```

### Made transition to regular download
* HASH_1
* HASH_3
* HASH_4
* HASH_6

#### Stayed as "Metadata download..."
* HASH_2
* HASH_5

**Comment by ferdnyc** [comment]
Realized I forgot to tag @parg: Repro logs and info posted!

**Comment by parg** [comment]
Thanks, please try B10 !

**Comment by ferdnyc** [comment]
@parg So far so good, I just double-clicked 9 magnets in RSSFeed Scanner and all of them successfully made the transition to real downloads, which almost certainly wouldn't have happened with previous betas. And no more "magnet returned: true" in the logs.

The real test will be seeing what happens to torrents that get added automatically by the plugin; it'll take a few days for enough new magnets to come over the feed to provide a large enough sample, so I'll leave this open a bit longer. But preliminary tests were all encouraging, I suspect this is sorted out now.

Thanks!

**Comment by parg** [comment]
We live in hope - thanks for the help debugging (so far)!

**Comment by parg** [comment]
Can you help this guy - https://github.com/BiglySoftware/BiglyBT/issues/2787 ? You seem to use RSS Feed Scanner :)

**Comment by ferdnyc** [comment]
@parg Heavily, yes. I'll try to drop some wisdom. 😉 

BTW, I _miiight_ have figured out what was going on here, and why I was seeing weird activation orderings you weren't.

Turns out (I just discovered, looking into my RSSFeed configs for the first time in months so that I can remember details for #2787) that _one_ of my RSSFeed filter rules was configured "Set state to: Queued" instead of "Set state to: Stopped" like I'd normally expect/intend. I don't know why I'd have set it differently, but maybe an instance of that "scrolling past a combobox scrolls the combobox" bug changed it on me, before that got fixed.

Regardless, the reason my torrents **weren't** being added Queued, despite that setting, is that I also have "By default add new torrents in a stopped state" checked in Options > Files > Torrents. So I guess we know which of those takes priority, now, although I guess when they interact it can be a bit racy.

Oops! 👼
